tests/test_example_meshes.cpp

```diff
@@ -16,6 +17,121 @@ struct MeshDebugInfo
+
+        std::queue<long> q;
+        q.push(fid);
+        while (!q.empty()) {
+            long cur_fid = q.front();
+            q.pop();
+
+            component_ids[cur_fid] = component_id;
+            auto f_tuple = m.tuple_from_face_id(cur_fid);
+            // push all adjacent faces
+            for (int j = 0; j < 3; ++j)
+            {
+                if (!m.is_boundary(f_tuple))
+                {   
+                    auto adj_face_tuple = m.switch_tuple(f_tuple, PrimitiveType::Face);
+                    long adj_fid = m.id(adj_face_tuple, PrimitiveType::Face);
+                    if (component_ids[adj_fid] == -1) {
+                        q.push(adj_fid);
+                    }  
+                }
+                f_tuple = m.switch_tuple(m.switch_tuple(f_tuple, PrimitiveType::Vertex), PrimitiveType::Edge);
+            } 
+        }
+        
+        component_id++;
+    }
+
+    return component_id;
+}
+
+int trimesh_simply_connected_components(const DEBUG_TriMesh& m)
+{
+    std::vector<int> component_ids;
+    return trimesh_simply_connected_components(m, component_ids);
+}
+
+int trimesh_n_bd_loops(const DEBUG_TriMesh& m)
+{
+    std::vector<int> bd_loops(m.capacity(PrimitiveType::Edge), -1);
+    auto all_edge_tuples = m.get_all(PrimitiveType::Edge);
+    int bd_loop_id = 0;
+    for (size_t i = 0; i < all_edge_tuples.size(); ++i) 
+    {
+        if (!m.is_boundary(all_edge_tuples[i])) {
+            continue;
+        }
+        long eid = m.id(all_edge_tuples[i], PrimitiveType::Edge);
+        if (bd_loops[eid] != -1) {
+            continue;
+        } // visited
+
+        Tuple cur_edge = all_edge_tuples[i];
+        long cur_eid = m.id(cur_edge, PrimitiveType::Edge);
+        // find one boundary loop
+        while (bd_loops[m.id(all_edge_tuples[i], PrimitiveType::Edge)] == -1 || cur_eid != m.id(all_edge_tuples[i], PrimitiveType::Edge)) 
+        {
+            bd_loops[cur_eid] = bd_loop_id;
+
+            // find next boundary edge
+            cur_edge = m.switch_tuple(cur_edge, PrimitiveType::Vertex);
+            cur_edge = m.switch_tuple(cur_edge, PrimitiveType::Edge);
+            while(!m.is_boundary(cur_edge))
+            {
+                cur_edge = m.switch_tuple(cur_edge, PrimitiveType::Face);
+                cur_edge = m.switch_tuple(cur_edge, PrimitiveType::Edge);
+            }
+            
+            cur_eid = m.id(cur_edge, PrimitiveType::Edge);
+        }
+        
+        bd_loop_id++;
+    }
+
+    return bd_loop_id;
+}
+
+int trimesh_euler_char(const DEBUG_TriMesh& m)
+{
+    auto simplex_counts = trimesh_simplex_counts(m);
+    int n_bd_loops = trimesh_n_bd_loops(m);
+    return simplex_counts[0] - simplex_counts[1] + simplex_counts[2] + n_bd_loops;
+}
+
+int trimesh_genus(const DEBUG_TriMesh& m)
+{
+    if (trimesh_simply_connected_components(m) != 1)
+    {
+        wmtk::logger().error("not a connected surface\n");
+        throw std::runtime_error("GenusComputeError");
```

Comment:
A minor thing but logging and throwing an error seems to me a bit overkill. I feel like throwing is enough. Anyways, the messages of the two should be the same I think.

Reply:
i agree but only if we have a top level try/catch that forwards exceptions to the logger.
there are cases where you want to send someone an output log (maybe a log from a file) and that won't contain the exceptions)..

on the other hand, the runtime error message is not practical (you're declaring it like you're specifying the runtime error is a class name, even though the aforementioned log message spits out more information. might as well make these two statements the same

Reply:
True, I didn't consider that. Then I guess having the logger and the error simultaneously with the same message might be the best option.

Reply:
alternative is to have a global catch that rethrows? i'm not sure how this works with debuggers though